inst/scribble/xaringanExtra-scribble.js

```diff
@@ -0,0 +1,490 @@
+/*
+ *  Scribble for remark.js/xaringan
+ *  Matthew T. Warkentin <warkentin@lunenfeld.ca>
+ */
+
+class Scribble {
+  constructor(opts) {
+    // User options
+    this.opts = opts
+    this.penColors = opts['pen_color'] || '#000000'
+    this.currColor = this.penColors[0]
+    this.penSize = opts['pen_size'] || 3
+    this.eraserSize = opts['eraser_size'] || 30
+    this.tolerance = this.eraserSize / 2
+    this.eraserColor = opts['eraser_color'] || 'rgba(0, 0, 0, 0.1)'
+    this.transparent = 'rgba(0, 0, 0, 0)'
+
+    // Fabric objects
+    this.fabrics
+    this.currFabric
+
+    // State objects
+    this.mouseDown = false
+    this.drawMode = false
+    this.eraseMode = false
+    this.hideToolbox = false
+    this.isPaused = false
+
+    this.eraserCursorMovement = this.eraserCursorMovement.bind(this)
+    this.eraser_impl = this.eraser_impl.bind(this)
+
+    // Toolbox objects
+    this.launchKey = 83 // "S" used to toggle toolbox viz
+    this.toolBox
+    this.btnWrapper
+    this.colWrapper
+    this.drawBtn
+    this.eraseBtn
+    this.clearBtn
+    this.colorPicker
+    this.eraserCursor
+
+    // Scribble initialization
+    this.addCanvasToAllSlides()
+    this.addCanvasHelpText()
+    this.assembleToolbox()
+    this.addToolboxToSlide()
+    this.initEraserCursor()
+    this.addToggleToolbox()
+    this.resizeContent()
+
+    this.addDrawing()
+    this.addClearing()
+    this.addErasing()
+
+    // Scribble continuous updating
+    slideshow.on('afterShowSlide', (slide) => {
```

Comment:
Rather than adding the buttons to each slide individually, I think we should put a single copy of the controls in the `.remark-slides-area` div. That way the controls are shared by all slides and the colors picked on one slide carry over to the next.

remark.search does this nicely and provides a [good pattern to follow](https://github.com/gadenbuie/xaringanExtra/blob/14fb86d8c829d98428e31c379724f0bf2a8d2027/inst/search/search.js#L38-L60). You have to use the slide scaler to position the controls and then update the position on resize. Another advantage is that the controls won't scale with the slides, which is a good thing here. We want the touch/click targets to be consistently sized regardless of slide size.

Reply:
Oh and you can use `slideshow.getCurrentSlideIndex()` to get the current slide to find the current canvas.

Reply:
This is already the current behaviour, for the most part. The toolbox is only present on the current visible slide. The toolbox gets passed from slide to slide when the slide changes (`slideshow.on('afterShowSlide', ...)`), so there is only a single copy of the controls and it lives in `.remark-slide-container outer-canvas-container`, so it is not subject to the remark slide scaler resizing. 

But I can see the advantage/simplicity of the toolbox just living at a global level, rather than the slide-local level. 

> That way the controls are shared by all slides and the colors picked on one slide carry over to the next.

I had actually added code to make the pen color fall back to the default choice on slide change, I thought this was preferable over carrying over color picker choices. What do you think?

Reply:
Oh I see. Yeah, I didn't realize the controls were moved rather than added to all slides. Hmm, I'm okay with leaving that alone and trying it for a bit. Since we're using the native color picker, which depends greatly on the OS and can range from easy to use to terribly frustrating, I think we should keep the same color from slide to slide. On MacOS it takes a lot of clicks to find the color you want (it's a lot easier on iPad) and it's super annoying to navigate those controls, esp. if you have to do it for each slide.